In [ ]:
import numpy as np
import pandas as pd
from matplotlib.pyplot import subplots
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import RidgeCV

CHAPTER 4: CLASSIFICATION

Problem 16: (Logistic Regression)

In [ ]:
Boston = pd.read_csv('/content/drive/MyDrive/Boston.csv')
print(Boston.head())

   Unnamed: 0     crim    zn  indus  chas    nox     rm   age     dis  rad  \
0           1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1   
1           2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2   
2           3  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2   
3           4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3   
4           5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3   

   tax  ptratio  lstat  medv  
0  296     15.3   4.98  24.0  
1  242     17.8   9.14  21.6  
2  242     17.8   4.03  34.7  
3  222     18.7   2.94  33.4  
4  222     18.7   5.33  36.2  


In [ ]:
Boston = Boston.drop(columns=['Unnamed: 0'])
print(Boston.head())

      crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

   lstat  medv  
0   4.98  24.0  
1   9.14  21.6  
2   4.03  34.7  
3   2.94  33.4  
4   5.33  36.2  


In [ ]:
Boston.columns

Index(['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'lstat', 'medv'],
      dtype='object')

In [ ]:
crim_above_med=np.where(Boston["crim"]>Boston["crim"].median(),1,0)
Boston['crim_above_med']=crim_above_med

In [ ]:
Boston['crim_above_med'].describe()

count    506.000000
mean       0.500000
std        0.500495
min        0.000000
25%        0.000000
50%        0.500000
75%        1.000000
max        1.000000
Name: crim_above_med, dtype: float64

In [ ]:
Boston.corr()

crim        zn     indus      chas       nox        rm  \
crim            1.000000 -0.200469  0.406583 -0.055892  0.420972 -0.219247   
zn             -0.200469  1.000000 -0.533828 -0.042697 -0.516604  0.311991   
indus           0.406583 -0.533828  1.000000  0.062938  0.763651 -0.391676   
chas           -0.055892 -0.042697  0.062938  1.000000  0.091203  0.091251   
nox             0.420972 -0.516604  0.763651  0.091203  1.000000 -0.302188   
rm             -0.219247  0.311991 -0.391676  0.091251 -0.302188  1.000000   
age             0.352734 -0.569537  0.644779  0.086518  0.731470 -0.240265   
dis            -0.379670  0.664408 -0.708027 -0.099176 -0.769230  0.205246   
rad             0.625505 -0.311948  0.595129 -0.007368  0.611441 -0.209847   
tax             0.582764 -0.314563  0.720760 -0.035587  0.668023 -0.292048   
ptratio         0.289946 -0.391679  0.383248 -0.121515  0.188933 -0.355501   
lstat           0.455621 -0.412995  0.603800 -0.053929  0.590879 -0.613808   
medv           -0.388305  0.360445 -0.483725  0.175260 -0.427321  0.695360   
crim_above_med  0.409395 -0.436151  0.603260  0.070097  0.723235 -0.156372   

                     age       dis       rad       tax   ptratio     lstat  \
crim            0.352734 -0.379670  0.625505  0.582764  0.289946  0.455621   
zn             -0.569537  0.664408 -0.311948 -0.314563 -0.391679 -0.412995   
indus           0.644779 -0.708027  0.595129  0.720760  0.383248  0.603800   
chas            0.086518 -0.099176 -0.007368 -0.035587 -0.121515 -0.053929   
nox             0.731470 -0.769230  0.611441  0.668023  0.188933  0.590879   
rm             -0.240265  0.205246 -0.209847 -0.292048 -0.355501 -0.613808   
age             1.000000 -0.747881  0.456022  0.506456  0.261515  0.602339   
dis            -0.747881  1.000000 -0.494588 -0.534432 -0.232471 -0.496996   
rad             0.456022 -0.494588  1.000000  0.910228  0.464741  0.488676   
tax             0.506456 -0.534432  0.910228  1.000000  0.460853  0.543993   
ptratio         0.261515 -0.232471  0.464741  0.460853  1.000000  0.374044   
lstat           0.602339 -0.496996  0.488676  0.543993  0.374044  1.000000   
medv           -0.376955  0.249929 -0.381626 -0.468536 -0.507787 -0.737663   
crim_above_med  0.613940 -0.616342  0.619786  0.608741  0.253568  0.453263   

                    medv  crim_above_med  
crim           -0.388305        0.409395  
zn              0.360445       -0.436151  
indus          -0.483725        0.603260  
chas            0.175260        0.070097  
nox            -0.427321        0.723235  
rm              0.695360       -0.156372  
age            -0.376955        0.613940  
dis             0.249929       -0.616342  
rad            -0.381626        0.619786  
tax            -0.468536        0.608741  
ptratio        -0.507787        0.253568  
lstat          -0.737663        0.453263  
medv            1.000000       -0.263017  
crim_above_med -0.263017        1.000000

In [ ]:
x=Boston[['indus','nox','age','dis','rad','tax']].values

y=Boston['crim_above_med'].values
(x_train,x_test,y_train,y_test)=train_test_split(x,y,random_state=1)

In [ ]:
X = Boston[['indus', 'nox', 'age', 'dis', 'rad', 'tax']]
X_const = sm.add_constant(X)
Y=(Boston['crim_above_med'] == 1).astype(int)
glm=sm.GLM(Y,X_const,family=sm.families.Binomial())
results = glm.fit()
print(results.summary())


                 Generalized Linear Model Regression Results                  
Dep. Variable:         crim_above_med   No. Observations:                  506
Model:                            GLM   Df Residuals:                      499
Model Family:                Binomial   Df Model:                            6
Link Function:                  Logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -122.38
Date:                Thu, 03 Oct 2024   Deviance:                       244.77
Time:                        21:03:14   Pearson chi2:                     294.
No. Iterations:                     9   Pseudo R-squ. (CS):             0.5945
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -24.5072      3.744     -6.545      0.0

In [ ]:
results.params

const   -24.507204
indus    -0.057598
nox      42.093845
age       0.013831
dis       0.206086
rad       0.598177
tax      -0.007260
dtype: float64

In [ ]:
probs = results.predict()
probs[:10]

array([0.14186054, 0.02613241, 0.02054682, 0.03597769, 0.04023172,
       0.04270494, 0.47420723, 0.59510774, 0.61447987, 0.5929735 ])

In [ ]:
labels = np.zeros(Boston.shape[0], dtype=int)
labels[probs > 0.5] = 1

conf_matrix = confusion_matrix(Boston['crim_above_med'], labels)

print(conf_matrix)

[[225  28]
 [ 40 213]]


In [ ]:
(225+213)/(225+213+28+40)

0.8656126482213439

In this case, logistic regression
correctly predicted crime rate 86.6% of the time.

CHAPTER 6 - LINEAR MODEL SELECTION AND REGULARIZATION

Problem 11: (LASSO/Ridge)

In [ ]:
# Set seed for reproducibility
np.random.seed(159)

boston = pd.read_csv('/content/drive/MyDrive/Boston.csv')

X = boston.drop(columns=['Unnamed: 0', 'crim'])
y = boston['crim']

In [ ]:
# Split data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Standardize the data for regularization methods
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [ ]:
lasso = Lasso(alpha=0.1)

# Fit model to training data
lasso.fit(X_train_scaled, y_train)

# Predictions and performance on test data
y_pred_lasso = lasso.predict(X_test_scaled)
lasso_mse = mean_squared_error(y_test, y_pred_lasso)
print('Lasso Test MSE:', lasso_mse)
print('Lasso Coefficients:', lasso.coef_)

Lasso Test MSE: 47.31442490271197
Lasso Coefficients: [ 0.68552554 -0.         -0.22260961 -0.62485611  0.14489728  0.10394735
 -1.41375269  4.50887328  0.         -0.36020619  0.07751059 -1.59727234]


Ridge Regression

In [ ]:
alphas = np.linspace(0.0001, 1, 100)
ridge_cv = RidgeCV(alphas=alphas, store_cv_results=True, scoring='neg_mean_squared_error')

# Fit model to training data
ridge_cv.fit(X_train_scaled, y_train)

# Print the best lambda (alpha) value and its corresponding MSE
print('Best lambda (alpha):', ridge_cv.alpha_)
print('Best cross-validated MSE:', -np.min(ridge_cv.cv_results_.mean(axis=0)))

Best lambda (alpha): 1.0
Best cross-validated MSE: 0.013249763833102
